src/csharp/Microsoft.Spark/Sql/ForeachWriter.cs

```diff
@@ -0,0 +1,191 @@
+        /// <returns>True if successful, false otherwise.</returns>
+        bool Open(long partitionId, long epochId);
+
+        /// <summary>
+        /// Called to process each <see cref="Row"/> in the executor side. This method
+        /// will be called only if <c>Open</c> returns <c>true</c>.
+        /// </summary>
+        /// <param name="row">The row to process.</param>
+        void Process(Row row);
+
+        /// <summary>
+        /// Called when stopping to process one partition of new data in the executor side. This is
+        /// guaranteed to be called either <see cref="Open(long, long)"/> returns <c>true</c> or
+        /// <c>false</c>. However, <see cref="Close(Exception)"/> won't be called in the following
+        /// cases:
+        /// <list type="bullet">
+        /// <item>
+        /// <description>
+        /// CLR/JVM crashes without throwing a <see cref="Exception"/>.
+        /// </description>
+        /// </item>
+        /// <item>
+        /// <description>
+        /// <see cref="Open(long, long)"/> throws an <see cref="Exception"/>.
+        /// </description>
+        /// </item>
+        /// </list>
+        /// </summary>
+        /// <param name="errorOrNull">
+        /// The <see cref="Exception"/> thrown during processing or null if there was no error.
+        /// </param>
+        void Close(Exception errorOrNull);
+    }
+
+    /// <summary>
+    /// Wraps a <see cref="IForeachWriter"/> and calls the appropriate methods as decribed in
+    /// the lifecycle documentation for the interface.
+    /// </summary>
+    internal class ForeachWriterWrapper
+    {
+        private readonly IForeachWriter _foreachWriter;
+
+        internal ForeachWriterWrapper(IForeachWriter foreachWriter) =>
+            _foreachWriter = foreachWriter;
+
+        internal IEnumerable<object> Execute(int partitionId, IEnumerable<Row> rows)
+        {
+            if (!TaskContextHolder.Get().LocalProperties.TryGetValue(
+                "streaming.sql.batchId",
+                out string epochIdStr) || !long.TryParse(epochIdStr, out long epochId))
+            {
+                throw new Exception(
+                    $"Could not get or parse batch id from TaskContext - batchId: {epochIdStr}");
+            }
+
+            Exception error = null;
+            bool opened = _foreachWriter.Open(partitionId, epochId);
+
+            try
+            {
+                if (opened)
+                {
+                    foreach (Row row in rows)
+                    {
+                        _foreachWriter.Process(row);
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                error = e;
+            }
+            finally
+            {
+                _foreachWriter.Close(error);
+
+                if (error != null)
+                {
+                    throw error;
+                }
+            }
+
+            return Enumerable.Empty<object>();
```

Comment:
Can we add a short comment why we are returning empty? (e.g., empty IEnumerable is returned because ForEach is a sink operation, but something needs to be returned to work within the UDF framework).